### ISSUE TITLE
Using waymo-dataset in keras

### ISSUE THREAD

**Issue by mjuergensen** [waymo-research/waymo-open-dataset#168]
Hey All,
I am trying to train a keras deep learning model with your data for my thesis. How ever i cant figure out how to preprocess the data to make it readable by the model.

I am loading the data with:
`preproc_data = tf.data.TFRecordDataset(data_path)` 

after preprocessing i want the data as Input for (or sth. similar):
`model_input = layers.Input(tensor=x_train_batch)`
`model_output = objdetect_layers(model_input)`

is there a way to process the given data to work with keras?
your tutorial is working just fine. How ever as mentioned before i could not figure out how to train a model with it. 
I would very much appreciate some help. thanks in advance.

best regards
Manuel

**Comment by peisun1115** [comment]
I think you can preprocess the data with another job to another format (e.g. in tensorflow.Example).

**Comment by SirJuergen** [comment]
Thanks for your fast answer!
I guess the thing Iam looking for is the parsing function. 
As far as i can see i need to use:
`return dataset.make_one_shot_iterator()` or sth. similar.
but for this i first need to map the features such as images, labels and bounding boxes. 
Is there a feature discription i could use?

thanks in advance.

**Comment by SirJuergen** [comment]
To say it more precisely:
I only need the images and the 2D bounding boxes. 
From the loaded tfrecord Data I would like to extract those features to pass it to a model.

```
def _parse_function(proto):
    feature = {
        "image": tf.VarLenFeature(),
        "label": tf.VarLenFeature()
    }

    parsed_features = tf.parse_single_example(proto, f)
    x = tf.reshape(parsed_features['x'] / 255, (IMG_SIZE[0], IMG_SIZE[1], 1))
    y = tf.cast(parsed_features['label'], tf.float32)

    return x, y

def load_dataset(input_path, batch_size, shuffle_buffer):`
    dataset = tf.data.TFRecordDataset(input_path)
    dataset = dataset.shuffle(shuffle_buffer).repeat()  # shuffle and repeat
    dataset = dataset.map(_parse_function, num_parallel_calls=16)
    dataset = dataset.batch(batch_size).prefetch(1)  # batch and prefetch

    return dataset.make_one_shot_iterator()

train_iterator = load_dataset(TRAIN_TFRECORDS, BATCH_SIZE, SHUFFLE_BUFFER)
val_iterator = load_dataset(VALIDATION_TFRECORDS, BATCH_SIZE, SHUFFLE_BUFFER)
```

Now i cant figure out how to build the feature dictionary fitting your data. 

thanks!

**Comment by amicoalbi** [comment]
@SirJuergen did you make any progress on that? I'm struggling with the features dictionary as well.

**Comment by SirJuergen** [comment]
for writing a parsing function like this helped:
`def parse_data(input_data):
"""
converts tfrecord data to readable arrays
Arguments:
input_data: frame from Waymo dataset
Output:
image_front_tensor: image as tensor
label_front: bounding boxes and class labels
"""
image_front_tensor = tf.image.decode_jpeg(input_data.images[0].image)
label_front = input_data.camera_labels[0].labels
return image_front_tensor, label_front`

afterwards i saved the data in a fitting format.

**Comment by amicoalbi** [comment]
Thank you!
